src/set.cr

```diff
@@ -496,10 +496,8 @@ struct Set(T)
 
   # Replaces every element of the set with a value returned from the block, and
   # returns `self`.
-  def map!(& : T -> T) : self
-    hash = @hash.transform_keys { |k, _| yield(k) }
-    @hash.clear
-    @hash.merge!(hash)
+  def map!(&block : T -> T) : self
+    @hash.transform_keys! { |k, _| block.call(k) }
     self
   end
```

Comment:
By some reason just `yield k` leads to a bit confusing compiling error:

```
In src/set.cr:500:36

 500 | @hash.transform_keys! { |k, _| yield(k) }
                                      ^
Error: can't use `yield` inside a proc literal or captured block
```

Reply:
Unrelated to this topic, but what do you find it confusing? Maybe we can improve it!
The error message explains what is wrong and for me this seems quite clear. 

It could of course be more helpful to guide towards understanding why, and how to move on (forward the captured block). I realize that the intricacies of blocks and procs may be a bit confusing when first being confronted with it.

Reply:
> but what do you find it confusing? 

This message sounds too generic - that you cannot use "yield" in **any** proc literal or block, what of course isn't true. So any hint that it's because of types mismatching would be helpful at least for newcomers.

Reply:
Ah, I see.

It's mentioned in the "Capturing blocks" that a captured block cannot contain "return" and "break". I suppose "yield" should be added into this list, shouldn't it? I am happy to open a PR with this minor change in the documentation.